extensions/amp-jwplayer/0.1/amp-jwplayer.js

```diff
@@ -94,50 +369,57 @@ class AmpJWPlayer extends AMP.BaseElement {
 
     this.contentSearch_ = element.getAttribute('data-content-search') || '';
     this.contentBackfill_ = element.getAttribute('data-content-backfill') || '';
+    this.contentContextual_ =
+      element.getAttribute('data-content-contextual') || '';
+    this.contentRecency_ = element.getAttribute('data-content-recency') || '';
+
+    installVideoManagerForDoc(this.element);
+    Services.videoManagerForDoc(this.element).register(this);
   }
 
   /** @override */
   layoutCallback() {
-    const iframe = this.element.ownerDocument.createElement('iframe');
-    const cid = encodeURIComponent(this.contentid_);
-    const pid = encodeURIComponent(this.playerid_);
     const queryParams = dict({
-      'search': this.getContextualVal() || undefined,
+      'search': this.getContextualVal_() || undefined,
       'contextual': this.contentContextual_ || undefined,
       'recency': this.contentRecency_ || undefined,
       'backfill': this.contentBackfill_ || undefined,
+      'isAMP': true,
     });
 
-    const baseUrl = `https://content.jwplatform.com/players/${cid}-${pid}.html`;
-    const src = addParamsToUrl(baseUrl, queryParams);
+    const url = this.getSingleLineEmbed_();
+    const src = addParamsToUrl(url, queryParams);
+    const frame = disableScrollingOnIframe(
+      createFrameFor(this, src, this.element.id)
+    );
 
-    iframe.setAttribute('frameborder', '0');
-    iframe.setAttribute('allowfullscreen', 'true');
-    iframe.src = src;
-    this.applyFillContent(iframe);
-    this.element.appendChild(iframe);
-    this.iframe_ = /** @type {HTMLIFrameElement} */ (iframe);
-    return this.loadPromise(iframe);
-  }
+    addUnsafeAllowAutoplay(frame);
+    disableScrollingOnIframe(frame);
+    // Subscribe to messages from player
+    this.unlistenFrame_ = listen(this.win, 'message', this.onMessage_);
+    // Forward fullscreen changes to player to update ui
+    this.unlistenFullscreen_ = listen(frame, 'fullscreenchange', () => {
```

Comment:
Curious: Why does this need to be handled here? If the frame is sending the event, why do we need to send it back?

Reply:
When iOS or safari, fullscreen is handled by our player, but otherwise we let AMP's utils handle it. In the case that someone triggers a fullscreen change via external UI, or if it's triggered by something else in the page, we need to tell the player the fullscreen state has changed in order to update the UI.